### ISSUE TITLE
static_mut_refs: Should the lint cover hidden references?

### ISSUE THREAD

**Issue by ehuss** [rust-lang/rust#123060]
There is some discussion on https://github.com/rust-lang/rust/issues/114447#issuecomment-1837156804 about whether or not [`static_mut_refs`](https://doc.rust-lang.org/nightly/rustc/lints/listing/warn-by-default.html#static-mut-refs) should be triggering on code that takes references without using the `&` sigil, such as method calls.

This should probably be resolved soon if this is to be part of the edition. If this needs to change after the edition, the implementation will likely require separate code paths for the lint versus edition behavior, which could become confusing, or risk unintentional breakage.

**Comment by traviscross** [comment]
@rustbot labels +I-lang-nominated

What we need to discuss is whether, e.g., these cases should trigger an error in Rust 2024:

```rust
// edition: 2024

static mut STATIC: &[u8; 3] = &[0, 1, 2];

fn main() { unsafe {
    let _ = STATIC.len();
    let _ = STATIC[0];
    let _ = format!("{:?}", STATIC);
}}
```

[Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&code=%2F%2F+edition%3A+2024%0A%0A%2F%2F+See%3A+https%3A%2F%2Fgithub.com%2Frust-lang%2Frust%2Fissues%2F123060%0A%0Astatic+mut+STATIC%3A+%26%5Bu8%3B+3%5D+%3D+%26%5B0%2C+1%2C+2%5D%3B%0A%0Afn+main%28%29+%7B+unsafe+%7B%0A++++let+_+%3D+STATIC.len%28%29%3B%0A++++let+_+%3D+STATIC%5B0%5D%3B%0A++++let+_+%3D+format%21%28%22%7B%3A%3F%7D%22%2C+STATIC%29%3B%0A%7D%7D%0A)

What was implemented in #117556 does not trigger an error for these.  However, @scottmcm expressed a view that these should be included also:

> I think this should be linting about anything that's created enough of a reference to trigger the reference rules -- if it can trigger UB when someone else holds a `&mut` to the static, that's worth linting about here.

cc @obeis

**Comment by Dirbaio** [comment]
+1 to this. I was quite confused when I saw that `MY_STATIC.some_mut_self_method()` doesn't generate a warning, since it's the same in the end as `&mut MY_STATIC`. I've also seen people in embedded bypassing the lint by rearranging code to get the reference by calling a `&mut self` method, without being aware it's "equally bad" as their previous code.

**Comment by tmandry** [comment]
Agreed that this should cover hidden references. Not sure if we need an FCP to decide it, but

@rustbot reviewed

**Comment by traviscross** [comment]
We reviewed this in the edition call today.  We'll close this when we merge #124895.

**Comment by shuangsilab** [comment]
I noticed that calling FnMut method on a static mut variable doesn't generate a warning. Is this behavior allowed by design?

**Comment by traviscross** [comment]
@shuangsilab: File that as a new issue if you would.
